infrastructure/src/main/java/org/corfudb/infrastructure/log/statetransfer/StateTransferManager.java

```diff
@@ -0,0 +1,293 @@
+         */
+        private final long startAddress;
+        /**
+         * End address of a segment range to transfer, inclusive and non-negative.
+         */
+        private final long endAddress;
+        /**
+         * A status of a transfer of a segment.
+         */
+        private final TransferSegmentStatus status;
+
+        @Override
+        public int compareTo(TransferSegment other) {
+            return Long.compare(this.getStartAddress(), other.getStartAddress());
+        }
+
+        /**
+         * Compute the total number of transferred addresses.
+         * {@link #endAddress} and {@link #startAddress} can only be non-negative longs such that
+         * {@link #endAddress} >= {@link #startAddress}.
+         *
+         * @return Sum of the total addresses transferred.
+         */
+        public long computeTotalTransferred() {
+            return endAddress - startAddress + 1L;
+        }
+
+        public static class TransferSegmentBuilder {
+
+            public void verify() {
+                if (startAddress < 0L || endAddress < 0L) {
+                    throw new IllegalStateException(
+                            String.format("Start: %s or end: %s " +
+                                    "can not be negative.", startAddress, endAddress));
+                }
+                if (startAddress > endAddress) {
+                    throw new IllegalStateException(
+                            String.format("Start: %s can not be " +
+                                    "greater than end: %s.", startAddress, endAddress));
+                }
+
+                if (status == null) {
+                    throw new IllegalStateException("Status should be defined.");
+                }
+            }
+
+            public TransferSegment build() {
+                verify();
+                return new TransferSegment(startAddress, endAddress, status);
+            }
+        }
+    }
+
+    /**
+     * A data class that represents a status of a segment to be transferred.
+     */
+    @Getter
+    @ToString
+    @EqualsAndHashCode
+    @Builder
+    public static class TransferSegmentStatus {
+        /**
+         * States of the segment:
+         * - {@link SegmentState#NOT_TRANSFERRED}: Segment is not transferred.
+         * - {@link SegmentState#TRANSFERRED}: Segment was transferred fully.
+         * - {@link SegmentState#RESTORED}: Segment was restored, and is present in the current layout.
+         * - {@link SegmentState#FAILED}: The state transfer of a segment has failed.
+         */
+        public enum SegmentState {
+            NOT_TRANSFERRED,
+            TRANSFERRED,
+            RESTORED,
+            FAILED
+        }
+
+        /**
+         * A state of a segment.
+         */
+        @Default
+        private final SegmentState segmentState = NOT_TRANSFERRED;
+        /**
+         * Total number of records transferred for this segment.
```

Comment:
What is a record?

I would suggest using "address" and denote order (i.e. sequential on the range) on the transfer.